sklearn/ensemble/voting_classifier.py

```diff
@@ -0,0 +1,227 @@
+
+        if self.voting not in ('soft', 'hard'):
+            raise ValueError("Voting must be 'soft' or 'hard'; got (voting=%r)"
+                             % voting)
+
+        if self.weights and len(self.weights) != len(self.estimators):
+            raise ValueError('Number of classifiers and weights must be equal'
+                             '; got %d weights, %d estimators'
+                             % (len(self.weights), len(self.estimators)))
+
+        self.le_ = LabelEncoder()
+        self.le_.fit(y)
+        self.classes_ = self.le_.classes_
+        self.estimators_ = []
+        for name, clf in self.estimators:
+            fitted_clf = clone(clf).fit(X, self.le_.transform(y))
+            self.estimators_.append(fitted_clf)
+        return self
+
+    def predict(self, X):
+        """ Predict class labels for X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        Returns
+        ----------
+        maj : array-like, shape = [n_samples]
+            Predicted class labels.
+        """
+        if self.voting == 'soft':
+            maj = np.argmax(self.predict_proba(X), axis=1)
+        else:  # 'hard' voting
+            predictions = self._predict(X)
+            maj = np.apply_along_axis(
+                                      lambda x:
+                                      np.argmax(np.bincount(x,
+                                                weights=self.weights)),
+                                      axis=1,
+                                      arr=predictions)
+
+        maj = self.le_.inverse_transform(maj)
+        return maj
+        
+    def _collect_probas(self, X):
+        """ Collect results from clf.predict calls. """
+        return np.asarray([clf.predict_proba(X) for clf in self.estimators_])
+
+    def _predict_proba(self, X):
+        """ Predict class probabilities for X in 'soft' voting """
+        avg = np.average(self._collect_probas(X), axis=0, weights=self.weights)
+        return avg
+        
+    def _check_proba(self):
+        if self.voting == 'hard':
+            raise AttributeError("predict_proba is not available when"
+                                 " voting=%r" % self.voting)    
+
+    @property
+    def predict_proba(self):
+        """ Compute probabilities of possible outcomes for samples in X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        Returns
+        ----------
+        avg : array-like, shape = [n_samples, n_classes]
+            Weighted average probability for each class per sample.
+        """
+        self._check_proba()
```

Comment:
Hm, yes, I also initially thought that this could be a little bit of an overkill ... but maybe it's slightly better this way in case there will be a more general check_proba function in some distant future